tools/lkl/lib/virtio.c

```diff
@@ -544,24 +546,79 @@ int virtio_dev_setup(struct virtio_dev *dev, int queues, int num_max)
 		return -LKL_ENOMEM;
 	}
 
-	avail = sizeof(lkl_virtio_devs) - (devs - lkl_virtio_devs);
-	num_bytes = snprintf(devs, avail, " virtio_mmio.device=%d@0x%lx:%d",
-			 mmio_size, (uintptr_t)dev->base, dev->irq);
-	if (num_bytes < 0 || num_bytes >= avail) {
-		lkl_put_irq(dev->irq, "virtio");
-		unregister_iomem(dev->base);
-		lkl_host_ops.mem_free(dev->queue);
-		return -LKL_ENOMEM;
+	if (!lkl_is_running()) {
+		avail = sizeof(lkl_virtio_devs) - (devs - lkl_virtio_devs);
+		num_bytes = snprintf(devs, avail, " virtio_mmio.device=%d@0x%lx:%d",
+				     mmio_size, (uintptr_t) dev->base, dev->irq);
+		if (num_bytes < 0 || num_bytes >= avail) {
+			lkl_put_irq(dev->irq, "virtio");
+			unregister_iomem(dev->base);
+			lkl_host_ops.mem_free(dev->queue);
+			return -LKL_ENOMEM;
+		}
+		devs += num_bytes;
+		dev->virtio_mmio_id = lkl_num_virtio_boot_devs++;
+	} else {
+		ret =
+		    lkl_sys_virtio_mmio_device_add((long)dev->base, mmio_size,
+						   dev->irq);
+		if (ret < 0) {
+			lkl_printf("can't register mmio device\n");
+			return -1;
+		}
+		dev->virtio_mmio_id = lkl_num_virtio_boot_devs + ret;
 	}
-	devs += num_bytes;
 
 	return 0;
 }
 
-void virtio_dev_cleanup(struct virtio_dev *dev)
+int virtio_dev_cleanup(struct virtio_dev *dev)
 {
+	char devname[100];
+	long fd, ret;
+
+	if (!lkl_is_running())
+		goto skip_unbind;
+
+	ret = lkl_mount_fs("sysfs");
+	if (ret < 0)
+		return ret;
+
+	if (dev->virtio_mmio_id >= virtio_get_num_bootdevs())
+		ret = snprintf(devname, sizeof(devname), "virtio-mmio.%d.auto",
+			       dev->virtio_mmio_id - virtio_get_num_bootdevs());
+	else
+		ret = snprintf(devname, sizeof(devname), "virtio-mmio.%d",
+			       dev->virtio_mmio_id);
+	if (ret < 0 || (size_t) ret >= sizeof(devname))
+		return -LKL_ENOMEM;
+
+	fd = lkl_sys_open("/sysfs/bus/platform/drivers/virtio-mmio/unbind",
+			  LKL_O_WRONLY, 0);
+	if (fd < 0)
+		return fd;
+
+	ret = lkl_sys_write(fd, devname, strlen(devname));
+	if (ret < 0)
+		return ret;
+
+	ret = lkl_sys_close(fd);
+	if (ret < 0)
+		return ret;
+
+	ret = lkl_sys_umount("/sysfs", 0);
```

Comment:
If sysfs is already mounted before this function, it will umount it. Can you umount only when lkl_mount_fs returns 0?